pkg/auth/handler/webapp/csrf_middleware.go

```diff
@@ -0,0 +1,204 @@
+package webapp
+
+import (
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/gorilla/csrf"
+	"github.com/gorilla/securecookie"
+	"github.com/sirupsen/logrus"
+
+	"github.com/authgear/authgear-server/pkg/lib/config"
+	"github.com/authgear/authgear-server/pkg/util/duration"
+	"github.com/authgear/authgear-server/pkg/util/httputil"
+	"github.com/authgear/authgear-server/pkg/util/jwkutil"
+	"github.com/authgear/authgear-server/pkg/util/log"
+
+	"github.com/authgear/authgear-server/pkg/auth/handler/webapp/viewmodels"
+	webapp "github.com/authgear/authgear-server/pkg/auth/webapp"
+)
+
+type CSRFMiddlewareLogger struct{ *log.Logger }
+
+func NewCSRFMiddlewareLogger(lf *log.Factory) CSRFMiddlewareLogger {
+	return CSRFMiddlewareLogger{lf.New("webapp-csrf-middleware")}
+}
+
+type CSRFMiddleware struct {
```

Comment:
Why do you duplicate this middleware here?

Reply:
Bec the original middleware is located in /pkg/auth/webapp and cannot import renderer and view model from /pkg/auth/handler/webapp (cause import cycle error), if we want to render the error page in middleware, I think we need to put csrf_middleware in pkg/auth/handler/webapp?